src/Microsoft.AspNetCore.BuildTools.ApiCheck/ApiListing/TypeDescriptor.cs

```diff
@@ -157,9 +157,37 @@ public static string GetTypeNameFor(TypeInfo type)
 
             if (type.IsGenericType)
             {
-                var name = type.GetGenericTypeDefinition().FullName;
-                typeName = name.Substring(0, name.IndexOf('`'));
-                typeName = $"{typeName}<{string.Join(", ", type.GetGenericArguments().Select(ga => GetTypeNameFor(ga.GetTypeInfo())))}>";
+                if (type.DeclaringType == null || type.DeclaringType.GetTypeInfo() == type)
+                {
+                    var name = type.GetGenericTypeDefinition().FullName;
+                    typeName = name.Substring(0, name.IndexOf('`'));
+                    typeName = $"{typeName}<{string.Join(", ", type.GetGenericArguments().Select(ga => GetTypeNameFor(ga.GetTypeInfo())))}>";
+                }
+                else
+                {
+                    var container = type.DeclaringType.GetTypeInfo();
+                    var prefix = GetTypeNameFor(container);
+                    var name = type.GetGenericTypeDefinition().FullName;
+                    var currentTypeGenericArguments = type.GetGenericTypeDefinition().GetGenericArguments()
+                    .Where(p => !container.GetGenericArguments().Any(cp => cp.Name == p.Name))
```

Comment:
Can you skip these instead. `.Skip(container.GetGenericArguments().Length)`? I think the arguments from the containing type always appear before the nested types.